Restore helper

rvdh opened the pull request.

Commit:
Add RestoreHelper

Commit:
Fix restoreIfNeeded test

Commit:
Work in progress

ChrisTerBeke commented:
I know you're still working on this, but my first impression is that a lot of this code is missing typing, docstrings and some other code style issues (e.g. 2 lines of space above classes). We don't have linting enabled in this repo yet, maybe we should do that to keep everything in line.

ChrisTerBeke commented:
You can also check https://github.com/Ultimaker/Meta for the exact codestyle rules.

ChrisTerBeke commented on README.md:
```diff
@@ -75,6 +75,7 @@ The following options are available to use in the `spec` section of the `yaml` c
 | `mongodb.replicas` | - | The amount of MongoDB replicas that should be available in the replica set. Must be an uneven positive integer and minimum 3. |
 | * `backups.cron` | - | The cron on which to create a backup to cloud storage.
 | * `backups.gcs.bucket` | - | The GCS bucket to upload the backup to. |
+| `backups.gcs.restore_bucket` | - | The GCS bucket that contains the backup we wish to restore. |
```
Maybe mention here that the default value is the `backups.gcs.bucket` configuration item?

ChrisTerBeke replied:
Also the `restore_from` entry is missing in the docs.

ChrisTerBeke commented on mongoOperator/helpers/MongoMonitoring.py:
```diff
@@ -0,0 +1,119 @@
+import logging
+
+
+class CommandLogger(monitoring.CommandListener):
```
Missing docstring.

ChrisTerBeke commented on mongoOperator/helpers/MongoMonitoring.py:
```diff
@@ -0,0 +1,119 @@
+
+class CommandLogger(monitoring.CommandListener):
+
+    def started(self, event):
```
Missing typing, docstring.

ChrisTerBeke commented on mongoOperator/helpers/MongoMonitoring.py:
```diff
@@ -0,0 +1,119 @@
+                      "{0.request_id} started on server "
+                      "{0.connection_id}".format(event))
+
+    def succeeded(self, event):
```
etc.

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+from mongoOperator.helpers.MongoResources import MongoResources
+from mongoOperator.models.V1MongoClusterConfiguration import V1MongoClusterConfiguration
+from mongoOperator.services.KubernetesService import KubernetesService
+# from mongoOperator.services.MongoService import MongoService
```
Remove commented line?

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+    RESTORE_RETRIES = 4
+    RESTORE_WAIT = 15.0
+
+    def __init__(self, kubernetes_service: KubernetesService):
```
__init__ functions need a -> None return type.

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+        """
+        self.kubernetes_service = kubernetes_service
+
+    def _getCredentials(self, cluster_object: V1MongoClusterConfiguration) -> dict:
```
The return type `dict` is too generic. It's better to use `Dict[str, any]`.

ChrisTerBeke commented on mongoOperator/helpers/MongoResources.py:
```diff
@@ -139,3 +91,15 @@ def _createReplicaConfig(cls, cluster_object: V1MongoClusterConfiguration) -> Di
         }
+
+    @classmethod
+    def getConnectionSeeds(cls, cluster_object: V1MongoClusterConfiguration) -> List[str]:
```
This method name is a bit weird for what it returns. Maybe we can call it `getMemberHostnames`.

ChrisTerBeke commented on mongoOperator/helpers/MongoMonitoring.py:
```diff
@@ -0,0 +1,119 @@
+                      "{0.topology_id}".format(event))
+
+
+class HeartbeatLogger(monitoring.ServerHeartbeatListener):
```
Our codestyle states that classes should be in their own file. We should create 4 files for these then. Alternatively we make a single monitoring class that can handle all the different event types.

ChrisTerBeke commented on mongoOperator/helpers/MongoMonitoring.py:
```diff
@@ -0,0 +1,119 @@
+
+
+class HeartbeatLogger(monitoring.ServerHeartbeatListener):
+    def __init__(self, cluster_object, all_hosts_ready_callback: Callable[[V1MongoClusterConfiguration], None]) -> None:
```
This method overrides init in a super class, so the super().__init__() should be called on the first line.

ChrisTerBeke commented on mongoOperator/helpers/MongoMonitoring.py:
```diff
@@ -0,0 +1,119 @@
+
+class TopologyLogger(monitoring.TopologyListener):
+
+    def __init__(self, cluster_object, replica_set_ready_callback: Callable[[V1MongoClusterConfiguration], None]) -> None:
```
This method overrides init in a super class, so the super().**init**() should be called on the first line.

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+        credentials_json = b64decode(credentials_encoded)
+        return json.loads(credentials_json)
+
+    def getLastBackup(self, cluster_object: V1MongoClusterConfiguration) -> str:
```
Maybe rename to getLastBackupStorageObjectName, as it doesn't actually return the backup itself?

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+        :param cluster_object: The cluster object from the YAML file.
+        :return: Whether a restore was executed or not.
+        """
+        if cluster_object.spec.backups.gcs.restore_from is not None:
```
We should use early return here to prevent nested code complexity:

```python
if cluster_object.spec.backups.gcs.restore_from is None:
    return False

backup_file = cluster_object.spec.backups.gcs.restore_from
if backup_file == "latest":
    backup_file = self.getLastBackup(cluster_object)

logging.info(...)

return True
```

This whey there is only 1 level of complexity in 1 lines.

ChrisTerBeke replied:
Shouldn't there be a default value for restore_from? As it's not required in the model.

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+        downloaded_file = self._downloadBackup(cluster_object, backup_file)
+
+        for _ in range(self.RESTORE_RETRIES):
+            # Wait for the replicaset to become ready
```
We can put this comment line above the for loop and remove the spaces between the for statement and the try statement.

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+                restore_output = check_output(["mongorestore", "--host", ','.join(hosts), "--gzip",
+                                               "--archive=" + downloaded_file])
+                logging.info("Restore output: %s", restore_output)
+                os.remove(downloaded_file)
```
This might throw some kind of OS error that currently is not caught.

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+        return self._downloadFile(
+            credentials=self._getCredentials(cluster_object),
+            bucket_name=cluster_object.spec.backups.gcs.restore_bucket \
+                    if cluster_object.spec.backups.gcs.restore_bucket \
```
This could be simplified to `bucket_name=cluster_object.spec.backups.gcs.restore_bucket or cluster_object.spec.backups.gcs.bucket`

ChrisTerBeke replied:
Or, to reduce complexity, make a variable from it first:

```python
bucket = cluster_object.spec.backups.gcs.restore_bucket or cluster_object.spec.backups.gcs.bucket
```

ChrisTerBeke commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,174 @@
+                              .format(backup_file, _, err.returncode, err.stderr, err.stdout))
+                sleep(self.RESTORE_WAIT)
+
+        raise SubprocessError("Could not restore '{}' after {} retries!".format(backup_file, self.RESTORE_RETRIES))
```
I think a TimeoutError would be more appropriate here.

ChrisTerBeke commented on restore-from-backup-local.sh:
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
```
Can we just use the existing local script and add an argument to also do a restore when using it?

ChrisTerBeke replied:
Now there's a lot of code duplication.

ChrisTerBeke commented on tests/fixtures/mongo_responses/initiate-ok.json:
```diff
@@ -0,0 +1 @@
+{"ok": 1.0, "operationTime": {"$timestamp": {"t": 1549963040, "i": 1}}, "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1549963040, "i": 1}}, "signature": {"hash": {"$binary": "AAAAAAAAAAAAAAAAAAAAAAAAAAA=", "$type": "00"}, "keyId": 0}}}
```
Format these responses with normal indentation so they're easier to read?

ChrisTerBeke commented on tests/helpers/TestAdminSecretChecker.py:
```diff
@@ -38,7 +38,7 @@ def test_createResource(self, b64encode_mock):
         self.assertEqual(self.kubernetes_service.createSecret.return_value, result)
         self.kubernetes_service.createSecret.assert_called_once_with(
-            self.secret_name, "default", {"username": "root", "password": "random-password"}
+            self.secret_name, self.cluster_object.metadata.namespace, {"username": "root", "password": "random-password"}
```
In tests it's better to use 'hardcoded' values for things like this, otherwise we fail to catch changes because the code/config change automatically propagates in the tests.

Commit:
Code cleanup

Commit:
Refactor

Commit:
Split command and server logger tests

Commit:
Fit strings

Commit:
Fit another string

Commit:
Some more code fixes

Commit:
Remove k8s stream events code as we don't use it due to instability of that feature

Commit:
Remove todo

Commit:
Remove unneeded brackets

Commit:
More cleanup in tests

Commit:
Remove old test code

Commit:
Reformat json example responses

Commit:
Some more code style, add TODO

Commit:
Remove unused import

Commit:
move some files around

rvdh committed:
Merge pull request #28 from Ultimaker/restore-2

Restore 2

Commit:
Add missing unittests, flake8 and pylint checks

ChrisTerBeke commented on Dockerfile:
```diff
@@ -25,6 +25,10 @@ ADD . .
 
+# Linting
+RUN flake8 . --count
+RUN pylint stardustAuthServerApi --rcfile=lib/stardustCommons/.pylintrc
```
This targets the wrong directory.

Commit:
Flake8 and pylint fixes

Commit:
Update documentation

Commit:
Utilize TimeoutError instead of SubprocessError for failing restores

Commit:
Use a single build-and-deploy script

Add example file for a specific backup to restore.
Only run the callback on the first host.
Increate initial command timeouts, so we give kubernetes a chance to
launch the new pods

Commit:
Rename test file

Catch OS error on attempting to remove file (continue anyway)
Call __init__ on superclasses

Commit:
Final processing of comments

Commit:
Check if the user exists before creating it

Commit:
Fix findUsers command and don't run when google_credentials are missing

Commit:
Delete secret before trying to create it

Commit:
Merge branch 'restore_helper' of https://github.com/Ultimaker/k8s-mongo-operator into restore_helper

Commit:
Fix creating the replicaSet when it's not necessary

Also remove unneed "|| true" statements and delete the google secret if
it already exists.

Commit:
Add missing docstrings

DanielSchiavini commented on Dockerfile:
```diff
@@ -2,31 +2,32 @@
+    echo "deb http://repo.mongodb.org/apt/debian stretch/mongodb-org/3.6 main" | tee /etc/apt/sources.list.d/mongodb-org-3.6.list && \
+    apt-get update && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y mongodb-org-tools mongodb-org-shell && \
+    pip install --no-cache-dir -r requirements.txt
```
Why do we need to reinstall all mongoDB tools when the requirements file changes?

rvdh replied:
The idea is to use as little RUN statements as possible, to have as little image layers as possible, to reduce the resulting image size. This will make pulling and starting the image quicker.

But I'm going to see if we can use an alpine image in this case, because the image is currently 1.2GB in size.

DanielSchiavini commented on examples/mongo-3-replicas-from-latest-backup.yaml:
```diff
@@ -0,0 +1,23 @@
+      # Set restore_from to 'latest' to use the last backup created when initializing the replicaset.
+      restore_from: "latest"
+      # set restore_bucket if the file in restore_from is in another bucket.
+      # restore_bucket:
```
this can be removed

ChrisTerBeke replied:
it's just here as an example for other users too.

DanielSchiavini commented on examples/mongo-3-replicas-from-specific-backup.yaml:
```diff
@@ -0,0 +1,23 @@
+      # Set restore_from to 'latest' to use the last backup created when initializing the replicaset.
+      restore_from: mongodb-backup-default-mongo-cluster-2019-02-07_132931.archive.gz
+      # set restore_bucket if the file in restore_from is in another bucket.
+      # restore_bucket:
```
this can be removed

DanielSchiavini commented on mongoOperator/ClusterManager.py:
```diff
@@ -0,0 +1,88 @@
+class ClusterChecker:
+    """ Manager that periodically checks the status of the MongoDB objects in the cluster. """
+
+    STREAM_REQUEST_TIMEOUT = (15.0, 5.0)  # connect, read timeout
```
I don't think this is used anymore

ChrisTerBeke replied:
indeed, removed it.

DanielSchiavini commented on mongoOperator/helpers/MongoResources.py:
```diff
@@ -28,101 +21,80 @@ def getMemberHostname(cls, pod_index, cluster_name, namespace) -> str:
         :return: The name of the host.
         """
-        return "{}-{}.{}.{}.svc.cluster.local".format(cluster_name, pod_index, cluster_name, namespace)
+        return "{cluster_name}-{}.{cluster_name}.{}.svc.cluster.local".format(pod_index, namespace,
```
Let's either use named parameters for all the vars

DanielSchiavini commented on mongoOperator/helpers/RestoreHelper.py:
```diff
@@ -0,0 +1,176 @@
+            return False
+
+        backup_file = cluster_object.spec.backups.gcs.restore_from
+        if backup_file == "latest":
```
let's move this to a constant (can be in the top of the class)

DanielSchiavini commented on mongoOperator/helpers/listeners/mongo/CommandLogger.py:
```diff
@@ -0,0 +1,34 @@
+from pymongo.monitoring import CommandStartedEvent, CommandListener, CommandSucceededEvent, CommandFailedEvent
+
+
+class CommandLogger(CommandListener):
```
Where's this used?

ChrisTerBeke replied:
Purely for having debug logging when running the operator.

DanielSchiavini commented on mongoOperator/helpers/listeners/mongo/HeartbeatListener.py:
```diff
@@ -0,0 +1,68 @@
+
+    def started(self, event: ServerHeartbeatStartedEvent) -> None:
+        """
+        When the heartbeat was sent.
```
I miss what this method is supposed to do and when..

DanielSchiavini commented on mongoOperator/helpers/listeners/mongo/HeartbeatListener.py:
```diff
@@ -0,0 +1,68 @@
+        When the heartbeat arrived.
+        :param event: The event.
+        """
+        # The reply.document attribute was added in PyMongo 3.4.
```
Is this relevant? I thought we'd only support the later versions..

DanielSchiavini commented on mongoOperator/helpers/listeners/mongo/HeartbeatListener.py:
```diff
@@ -0,0 +1,68 @@
+        super().__init__()
+        self._cluster_object: V1MongoClusterConfiguration = cluster_object
+        self._expected_host_count: int = cluster_object.spec.mongodb.replicas
+        self._hosts: Dict[str, int] = {}
```
What is this dict supposed to contain?

rvdh replied:
A key (hostname of a cluster node) and a value (1) indicating the host is ready.

DanielSchiavini commented on mongoOperator/helpers/listeners/mongo/HeartbeatListener.py:
```diff
@@ -0,0 +1,68 @@
+        """
+        # The reply.document attribute was added in PyMongo 3.4.
+        logging.debug("Heartbeat to server %s succeeded with reply %s", event.connection_id, event.reply.document)
+        self._hosts[event.connection_id] = 1
```
What does 1 mean in this context?

rvdh replied:
That the host is up and running.

DanielSchiavini commented on mongoOperator/helpers/resourceCheckers/BaseResourceChecker.py:
```diff
@@ -11,14 +11,15 @@
     Base class for services that can check Kubernetes resources.
     """
 
     # this is the resource type, e.g. V1Service or V1StatefulSet.
```
Comment was left behind

DanielSchiavini commented on mongoOperator/services/KubernetesService.py:
```diff
@@ -103,21 +102,21 @@ def getMongoObject(self, name: str, namespace: str) -> V1MongoClusterConfigurati
                                                                     name)
 
-    def listAllServicesWithLabels(self, labels: Dict[str, str] = DEFAULT_LABELS) -> V1ServiceList:
+    def listAllServicesWithLabels(self, labels: Dict[str, str] = None) -> V1ServiceList:
```
Should be `Optional[Dict[...]]` if the default is `None`. Same below..

DanielSchiavini reviewed (commented):
Great improvements in this PR. I couldn't review the tests yet, sorry!

DanielSchiavini commented on mongoOperator/models/BaseModel.py:
```diff
@@ -55,7 +55,7 @@ def __eq__(self, other: any) -> bool:
         :return: True if they are equal, False otherwise.
         """
-        return type(other) == type(self) and other.to_dict() == self.to_dict()
+        return isinstance(other, type(self)) and other.to_dict() == self.to_dict()
```
Shouldn't the types be exactly the same?

rvdh replied:
The previous check resulted in:
mongoOperator/models/BaseModel.py:58:15: C0123: Using type() instead of isinstance() for a typecheck. (unidiomatic-typecheck)

DanielSchiavini commented on tests/TestMongoOperator.py:
```diff
@@ -14,11 +14,28 @@ class TestMongoOperator(TestCase):
     @patch("mongoOperator.MongoOperator.sleep")
     @patch("mongoOperator.MongoOperator.ClusterChecker")
     def test_run(self, checker_mock, sleep_mock):
+        checker_mock.return_value.collectGarbage.side_effect = None, Exception()  # break the 3rd run
```
The 3rd?

DanielSchiavini commented on tests/helpers/TestBackupChecker.py:
```diff
@@ -3,6 +3,7 @@
 # -*- coding: utf-8 -*-
 import json
 from base64 import b64encode
+from typing import cast
```
This file should be renamed to TestBackupHelper, but tbh Helper doesn't say anything about what the class should do..

rvdh replied:
Maybe BackupManager is a better suited name?

DanielSchiavini commented on tests/helpers/TestClusterChecker.py:
```diff
@@ -43,109 +43,58 @@ def test_checkExistingClusters_empty(self):
 
-    @patch("mongoOperator.helpers.ClusterChecker.ClusterChecker.checkCluster")
-    @patch("kubernetes.watch.watch.Watch.stream")
-    def test_streamEvents_add_update(self, stream_mock, check_mock):
```
Don't we support these events anymore?

ChrisTerBeke replied:
Kubernetes event stream was never used (and now removed) because it seemed unstable when doing tests with it. For example there were many duplicate events.

DanielSchiavini commented on tests/helpers/TestCommandLogger.py:
```diff
@@ -0,0 +1,30 @@
+    command_logger = CommandLogger()
+
+    def test_started(self):
+        self.command_logger.started(event = cast(CommandStartedEvent, CommandEventMock()))
```
Is the cast necessary? A Mock instance doesn't usually need to be casted. Same in the other tests.

ChrisTerBeke replied:
It's not really needed to run the tests, but PyCharm is showing it as warning, so it's more a prettiness thing.

DanielSchiavini replied:
IMO pycharm needs to be configured better, cause this makes the code much less readable; besides it's a lot of nonproductive work. My PyCharm is not doing this, can we check out if it's possible to disable only for mock objects?

DanielSchiavini commented on tests/helpers/TestCommandLogger.py:
```diff
@@ -0,0 +1,30 @@
+    duration_micros = 10000
+
+
+class TestCommandLogger(TestCase):
```
There are zero assertions in this test

DanielSchiavini commented on tests/helpers/TestHeartBeatListener.py:
```diff
@@ -0,0 +1,71 @@
+
+        heartbeat_logger.succeeded(event=cast(ServerHeartbeatSucceededEvent, heartbeat_event_mock))
+
+        self._onAllHostsReadyCallback.assert_called_once_with(self.cluster_object)
```
With this many calls, why is there only one assertion?

DanielSchiavini commented on tests/helpers/TestHeartBeatListener.py:
```diff
@@ -0,0 +1,71 @@
+
+        heartbeat_logger.started(event=cast(ServerHeartbeatStartedEvent, Mock(spec=ServerHeartbeatStartedEvent)))
+
+    def test_succeeded(self):
```
This test will pass if the callback is called in the first succeeded() call

DanielSchiavini commented on tests/helpers/TestHeartBeatListener.py:
```diff
@@ -0,0 +1,71 @@
+        heartbeat_event_mock.reply.document = {"info": "Does not have a valid replica set config"}
+        heartbeat_logger.succeeded(event=cast(ServerHeartbeatSucceededEvent, heartbeat_event_mock))
+
+    def test_succeeded_already_called(self):
```
This is not testing much, it's only making sure the coverage is reached

DanielSchiavini commented on tests/helpers/TestRestoreHelper.py:
```diff
@@ -0,0 +1,142 @@
+        restore_mock.reset_mock()
+
+        self.cluster_dict = getExampleClusterDefinition()
+        print(repr(self.cluster_dict))
```
Please remove print statements

DanielSchiavini commented on tests/helpers/TestRestoreHelper.py:
```diff
@@ -0,0 +1,142 @@
+
+        self.restore_helper.restoreIfNeeded(self.cluster_object)
+        print(restore_mock.mock_calls)
+        assert not restore_mock.called, "restore_mock should not have been called"
```
please use `self.assertFalse(restore_mock.called, "...")`

DanielSchiavini commented on tests/helpers/TestRestoreHelper.py:
```diff
@@ -0,0 +1,142 @@
+
+        self.assertEqual("Could not restore '" + expected_backup_name + "' after 4 retries!", str(context.exception))
+        self.assertEqual(4, subprocess_mock.call_count)
+        # TODO: assert calls on unused mocks
```
Still TODO's in this file

DanielSchiavini commented on tests/helpers/TestRestoreHelper.py:
```diff
@@ -0,0 +1,142 @@
+    @patch("mongoOperator.helpers.RestoreHelper.StorageClient")
+    @patch("mongoOperator.helpers.RestoreHelper.ServiceCredentials")
+    @patch("mongoOperator.helpers.RestoreHelper.check_output")
+    def test_restore_mongo_error(self, subprocess_mock, gcs_service_mock, storage_mock, os_mock):
```
if we are mocking stuff we should check how the mock gets called

DanielSchiavini commented on tests/services/TestMongoService.py:
```diff
@@ -40,263 +47,204 @@ def setUp(self):
-                }
-            }
-        }
+        self.initiate_ok_response = loads("""
```
can't we simply put the dict in here then?

rvdh replied:
The 'loads' method of bson.json_util converts $timestamp and other keys like that to something actually json. If that "loads" method changes in the future, the test is going to fail because we're not expecting that change. But that could be exactly what you want.. I don't know?

Commit:
Small changes to make it work on MacOS with Docker for Mac

Commit:
Remove casting in tests

Commit:
Fix most review comments

Commit:
Process comments

Commit:
Rename google_credentials.json to google-credentials.json

Commit:
Add assertions and remove unneeded test
